### ISSUE TITLE
about memory

### ISSUE THREAD

**Issue by leoDreamer** [CacheControl/json-rules-engine#132]
Hi, dear author
I have plan to use this library in my project, but there are some questions.
First I add 1000 rules in engine it spend 25MB memory, then start to add facts(with option cache false), the memory increase with the fact add, when add 1000 facts the memory spend 1304MB, there some method to disable memory increase ?
+ this is my test code:
```javascript
const Engine = require('json-rules-engine').Engine
const Rule = require('json-rules-engine').Rule

const engine = new Engine(undefined, {
  allowUndefinedFacts: true
})

const opt = {
  rule: 1000
}

function getRandor (num) {
  return parseInt((Math.random() * 1000000) % num) + 1
}

function mulitFact () {
  const len = opt.rule
  for(var i = 0; i < len; i++) {
    const rule = new Rule({
      conditions: {
        all: [{
            fact: `s${i}`,
            operator: 'equal',
            value: true
          },{
            fact: `s${i + 1}`,
            operator: 'equal',
            value: true
          }]
      },
      event: {
        type: `scence${i}-emit`
      }
    })
    engine.addRule(rule)
    console.log(`add rule ${i + 1} ${(process.memoryUsage().rss / 1024 / 1024).toFixed(2)}MB`)
  }
}

function run (id, value) {
  engine.addFact(id, value, { cache: false })
  return  engine.run()
    .then(resp => {
      console.log(`run reslut: ${resp.length}`)
      // engine.removeFact(id)
    })
}

(async function main () {
  mulitFact()

  for (var i = 0; i < 10000; i++) {
    const id = `s${getRandor(opt.rule)}`
    console.time(id)
    await run(id, true)
    console.timeEnd(id)
    console.log(`add fact ${i} - ${id} ${(process.memoryUsage().rss / 1024 / 1024).toFixed(2)}MB`)
  }
})()
```
By the way, `removeFact` after run also lead to memory increase.

**Comment by CacheControl** [comment]
Hey @leoDreamer, thanks for raising this. Most users of this library use dozens/hundreds of rules, so this is on the rarer side of our use cases.

Would you mind trying something for me? Can you change the `event.type` to a constant string and use a [custom parameter](https://github.com/CacheControl/json-rules-engine/blob/master/docs/rules.md#constructorobject-optionsstring-json) to uniquely identify which rule was triggered? e.g.:

```js
event: {
    type: 'scence$',
    params: {
      i
    }
  }
```
Every unique event registers an event emitter, and I'm wondering if that is impacting the memory usage. Let me know how it goes!

**Comment by leoDreamer** [comment]
Hi, it seems not work that change the `event.type` to a constant string, I'm doing some debug, the main reason like **UNDEFINED_FACT**, this is a memory heapdump printscreen.
![image](https://user-images.githubusercontent.com/16728066/57674627-595a5980-7652-11e9-935b-15c64c6b3e79.png)

**Comment by CacheControl** [comment]
@leoDreamer I see, did you confirm the memory issue goes away when you turn off allowing undefined facts?

**Comment by leoDreamer** [comment]
I'm not sure, because the test code can't run. it's similar with #96，change code to `return Promise.resolve(undefined)` the memory leak resolved.
https://github.com/CacheControl/json-rules-engine/blob/557d8ef485a723b6218e4b1ddce006b6a1c49ea1/src/almanac.js#L87-L89

**Comment by CacheControl** [comment]
@leoDreamer I was able to use your code above to reproduce the issue. I've published **v3.0.0**. There is still a very small increase in memory (a few kb) with each run, however nowhere near the multiple MBs from before.

Let me know if things haven't improved on your end.
